feat: Automatically reset DNS-configured connections on DNS change

hessjcg opened the pull request:
When a connection is configured using a DNS domain name, the connector will poll the DNS TXT record every 30 seconds. If the value of the DNS record changes, the connector will close all connections to the old instance, and direct new connections to the updated instance. 

Part of #842

enocom commented on dialer.go:
```diff
@@ -516,12 +552,94 @@ func (d *Dialer) Close() error {
+	}
+
+	// List all the cache entries created with a domain name
+	d.lock.RLock()
```
What does this lock protect? Are we sure we want to have it protect cache entries as well?

hessjcg replied:
`d.lock` (which was here before this PR) the contents of `d.cache`. Before this PR, the Dialer only added or removed single keys. Iterating through the whole cache is new behavior required by periodic DNS reconciliation.

`d.lock` only needs to protect the state of the `d.cache`. There is a separate lock on each cache entry `entry.cache.lock` which protects the internal state of each `connectionInfoCache.`

enocom commented on dialer.go:
```diff
@@ -516,12 +552,94 @@ func (d *Dialer) Close() error {
+
+	for _, entry := range cacheEntries {
+		// Remove closed sockets from cache.openSockets
+		entry.cache.lock.Lock()
```
`mu` is a more typical name in Go and avoids the stutter: lock.Lock()

hessjcg replied:
I renamed the mutex fields to `mu`.

hessjcg commented on README.md:
```diff
@@ -292,6 +293,40 @@ func connect() {
+### Automatic fail-over using DNS domain names
+
+When you configure the connector using a domain name, the connector will 
+periodically check if the DNS record for an instance changes. When connector 
```
Fixed.

hessjcg commented on README.md:
```diff
@@ -292,6 +293,40 @@ func connect() {
+
+The connector will automatically close all existing connections to
+`my-project:region:my-instance`. This may cause errors in your application as
+database queries in progress will fail.
```
Reworded.

hessjcg commented on dialer.go:
```diff
@@ -275,8 +285,23 @@ func NewDialer(ctx context.Context, opts ...Option) (*Dialer, error) {
+func (d *Dialer) initFailoverRoutine(ctx context.Context, p time.Duration) {
+	d.domainNameTicker = time.NewTicker(p)
+	go func() {
+		for range d.domainNameTicker.C {
```
Fixed.

hessjcg commented on dialer.go:
```diff
@@ -516,12 +552,94 @@ func (d *Dialer) Close() error {
 	return nil
 }
 
+func (d *Dialer) pollDomainNames(ctx context.Context) {
```
Yeah, this is a little odd. Dialer's primary thing is to establish connections. But as part of establishing a connection for DNS, it also manages the cache used to lookup ConnectionInfoCache instances. Checking for DNS changes falls under "managing the cache". 

In the Java connector, we separated these responsibilities: `InternalConnectorRegistry` does "manage the cache" while `Connector` does "establish connections." 

Maybe we should split `Dialer` in a similar way.

enocom replied:
Looking at this code some more, I think monitoredCache is pretty much what we're looking for. The challenge is establishing a clean interface between the dialer (which tracks all the caches) and the monitored cache itself.

I see a few places below where you could move code into methods on monitoredCache.

hessjcg commented on dialer.go:
```diff
@@ -516,12 +552,94 @@ func (d *Dialer) Close() error {
+					entry.cn.DomainName(), entry.cn.String(), newCn.String()))
+			// preload the new cache entry
+			b := entry.cache.UseIAMAuthN()
+			d.connectionInfoCache(ctx, newCn, &b)
```
Yeah, maybe we need pull the "create new cache entry" logic out of `d.connectionInfoCache()` into its own method.

hessjcg reviewed (commented):
I updated the code with some refactoring based on Eno's comments. Please take another look.

enocom commented on dialer.go:
```diff
@@ -538,62 +580,168 @@ func (d *Dialer) Close() error {
+	for _, entry := range cacheEntries {
+		// Remove closed sockets from cache.openSockets
+		entry.cache.mu.Lock()
+		var newOpenSockets []*instrumentedConn
```
nit: this could just be open. The type communicates what open is.

enocom commented on dialer.go:
```diff
@@ -520,6 +561,7 @@ func (i *instrumentedConn) Write(b []byte) (int, error) {
 // Close delegates to the underlying net.Conn interface and reports the close
 // to the provided closeFunc only when Close returns no error.
 func (i *instrumentedConn) Close() error {
+	i.closed = true
```
This isn't thread safe since we have a read happening below on this value.

enocom replied:
Instead of tracking closed (which requires more locking), could we just always close the connection and ignore double closed errors?

enocom commented on dialer.go:
```diff
@@ -538,62 +580,168 @@ func (d *Dialer) Close() error {
+
+		// the domain name no longer resolves to a valid instance
+		if err != nil {
+			d.logger.Debugf(ctx, "[failover] unable to resolve DNS for instance %s: %v", entry.cn.DomainName(), err)
```
We should have a `continue` after this log statement.

enocom commented on dialer.go:
```diff
@@ -115,15 +116,19 @@ type connectionInfoCache interface {
 	openConns *uint64
 
+	mu          sync.Mutex
+	openSockets []*instrumentedConn
```
The underlying Go interface is net.Conn. Let's use the same abstraction and call this `openConns`.

enocom commented on dialer.go:
```diff
@@ -538,62 +580,168 @@ func (d *Dialer) Close() error {
+	// List all the cache entries created with a domain name
+	d.mu.RLock()
+	cacheEntries := make([]cacheEntry, 0, len(d.cache))
+	for cn, cache := range d.cache {
```
Looks like we're copying all the keys and values out of this map into a slice. To avoid introducing a new type (cacheEntry), why don't we just allocate a new map of cn -> cache and then iterate over its keys and values? This reduces some of the complexity that adding a new type brings.

enocom commented on dialer.go:
```diff
@@ -538,62 +580,168 @@ func (d *Dialer) Close() error {
+func (d *Dialer) closeDomainNameChanged(ctx context.Context, cn instance.ConnName, cache monitoredCache, err error) {
+	d.removeCached(ctx, cn, cache, err)
+	if atomic.LoadUint64(cache.openConns) > 0 {
+		for _, socket := range cache.openSockets {
```
Shouldn't we also reset openSockets here too?

enocom reviewed (commented):
This looks better, but I think there are still some interface problems. I've left detailed comments below.

hessjcg reviewed (commented):
I responded to comments. I also significantly restructured the code. Please review #865 first, and then return to this one.

hessjcg commented:
@enocom Please review #865 before you take another look at this one.

enocom commented:
To summarize our conversation:

1. We'll move the ticker onto the monitoredCache
2. The monitoredCache will need to have some state when the DNS changes
3. New calls to get a connectionInfoCache will check for that state to see if the dialer needs to throw out the existing cache

hessjcg commented:
Ok, I've updated the code to reflect our discussion.


I this as discussed: I removed the "lock the world" operations. I moved the domain name ticker to monitoredCache. I made the cache get operation check if an entry is closed.

I still think is valuable to separate the concerns: dialing connections, maintaining a consistent and thread-safe cache state into two classes. `Dialer` should do the dialing, `dialerCache` should maintain cache thread safety and implement the logic around keeping the cache consistent when there are entries with the same domain name but different instance names, or when there are entries that were closed and need to be replaced.

hessjcg commented:
This is superseded by by #868
